quickjs.c

```diff
@@ -9198,6 +9248,62 @@ static int js_shape_prepare_update(JSContext *ctx, JSObject *p,
     return 0;
 }
 
+int js_shape_delete_watchpoints(JSRuntime *rt, JSShape *shape, intptr_t target) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+        if (o->delete_callback)
+            if (!o->delete_callback(rt, (intptr_t)o->ref, o->offset, o->atom, target)) {
+                list_del(el);
+                js_free_rt(rt, o);
+                break;
+            }
+    }
+end:
+    return 0;
+}
+
+int js_shape_free_watchpoints(JSRuntime *rt, JSShape *shape) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+        if (o->free_callback)
+            o->free_callback(rt, (intptr_t)o->ref, o->offset, o->atom);
+        list_del(el);
+        js_free_rt(rt, o);
+    }
+    list_empty(shape->watchpoint);
+    js_free_rt(rt, shape->watchpoint);
+end:
+    return 0;
+}
+
+JSInlineCacheWatchpoint *js_shape_create_watchpoint(JSRuntime *rt, JSShape *shape, intptr_t ptr, uint8_t offset, JSAtom atom,
+                                                    watchpoint_delete_callback *delete_callback,
+                                                    watchpoint_free_callback *free_callback)
+{
+    JSInlineCacheWatchpoint *o;
+    o = (JSInlineCacheWatchpoint *)js_malloc_rt(rt, sizeof(JSInlineCacheWatchpoint));
```

Comment:
Superfluous cast:
```suggestion
    o = js_malloc_rt(rt, sizeof(JSInlineCacheWatchpoint));
```
And while it's not exactly wrong or unidiomatic to use the type name, I'd write it as this:
```suggestion
    o = js_malloc_rt(rt, sizeof(*o));
```
That way it's automatically correct when o's type changes.